File: frontend/src/api/riskScores.ts
import { api } from './client';

export interface RiskScore {
  employee_id: string;
  id: string;
  score: number;
  risk_level: string;
  breakdown: Record<string, unknown>;
  calculated_at: string;
}

export async function listRiskScores(params?: {
  employee_id?: string;
  risk_level?: string;
  limit?: number;
}): Promise<RiskScore[]> {
  const query = new URLSearchParams();
  if (params?.employee_id) query.set('employee_id', params.employee_id);
  if (params?.risk_level) query.set('risk_level', params.risk_level);
  if (params?.limit) query.set('limit', String(params.limit));
  const qs = query.toString();
  return api.get<RiskScore[]>(`/risk-scores${qs ? `?${qs}` : ''}`);
}

export async function getRiskDistribution(): Promise<Record<string, number>> {
  return api.get<Record<string, number>>('/risk-scores/distribution');
}

export async function getRiskScoreHistory(employeeId: string, limit = 100): Promise<RiskScore[]> {
  return api.get<RiskScore[]>(`/risk-scores/history/${employeeId}?limit=${limit}`);
}

export interface RiskTrendPoint {
  date: string;
  average_score: number;
}

export interface DepartmentRisk {
  department: string;
  employees: number;
  average_score: number;
  max_score: number;
  high_risk_count: number;
}

export interface RiskAnalytics {
  average_score: number;
  max_score: number;
  min_score: number;
  total_employees_scored: number;
  distribution: Record<string, number>;
  trend: RiskTrendPoint[];
  department_breakdown: DepartmentRisk[];
  top_contributors: {
    id: string;
    name: string;
    department: string | null;
    designation: string | null;
    risk_score: number;
    risk_level: string;
    breakdown: Record<string, unknown>;
  }[];
}

export interface RiskScoreCalculateResult {
  calculated: number;
  days: number;
  average_score: number;
  distribution: Record<string, number>;
  message: string;
}

export async function calculateRiskScores(days = 30): Promise<RiskScoreCalculateResult> {
  return api.post<RiskScoreCalculateResult>(`/risk-scores/calculate?days=${days}`);
}

export async function getRiskAnalytics(days = 30): Promise<RiskAnalytics> {
  return api.get<RiskAnalytics>(`/risk-scores/analytics?days=${days}`);
}

File: app/api/v1/risk_scores.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import Optional
import uuid

from app.db.base import get_db
from app.models.risk_score import RiskScore
from app.models.employee import Employee
from app.schemas.risk_score import (
    RiskScoreOut,
    RiskScoreCalculateResult,
    RiskAnalytics,
)
from app.services.risk_scoring import (
    calculate_risk_scores,
    get_risk_analytics,
)
from app.core.deps import require_role, require_staff
from app.models.user import User

router = APIRouter(prefix="/api/v1/risk-scores", tags=["Risk Scores"])


@router.get("", response_model=list[RiskScoreOut])
def list_risk_scores(
    employee_id: Optional[uuid.UUID] = Query(None),
    risk_level: Optional[str] = Query(None),
    limit: int = Query(100, ge=1, le=500),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_staff),
):
    subquery = (
        db.query(
            RiskScore.employee_id,
            RiskScore.id,
            RiskScore.score,
            RiskScore.risk_level,
            RiskScore.breakdown,
            RiskScore.calculated_at,
        )
        .distinct(RiskScore.employee_id)
        .order_by(RiskScore.employee_id, RiskScore.calculated_at.desc())
    ).subquery()

    query = db.query(subquery)
    if employee_id:
        query = query.filter(subquery.c.employee_id == employee_id)
    if risk_level:
        query = query.filter(subquery.c.risk_level == risk_level)
    return query.limit(limit).all()


@router.get("/history/{employee_id}", response_model=list[RiskScoreOut])
def get_risk_score_history(
    employee_id: uuid.UUID,
    limit: int = Query(100, ge=1, le=500),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_staff),
):
    return (
        db.query(RiskScore)
        .filter(RiskScore.employee_id == employee_id)
        .order_by(RiskScore.calculated_at.desc())
        .limit(limit)
        .all()
    )


@router.get("/distribution")
def get_risk_distribution(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_staff),
):
    from sqlalchemy import func

    subquery = (
        db.query(
            RiskScore.employee_id,
            RiskScore.risk_level,
        )
        .distinct(RiskScore.employee_id)
        .order_by(RiskScore.employee_id, RiskScore.calculated_at.desc())
    ).subquery()

    distribution = (
        db.query(subquery.c.risk_level, func.count())
        .group_by(subquery.c.risk_level)
        .all()
    )
    return {level: count for level, count in distribution}


@router.post("/calculate", response_model=RiskScoreCalculateResult)
def recalculate_risk_scores(
    days: int = Query(30, ge=1, le=365),
    employee_id: Optional[uuid.UUID] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("administrator", "security_manager", "soc_engineer")),
):
    """
    Run the insider risk scoring engine and persist fresh scores.

    Recomputes the weighted threat models for every employee (or a single
    one) and stores the results as new RiskScore records.
    """
    result = calculate_risk_scores(
        db, days=days, employee_id=str(employee_id) if employee_id else None
    )
    return RiskScoreCalculateResult(**result)


@router.get("/analytics", response_model=RiskAnalytics)
def risk_analytics(
    days: int = Query(30, ge=1, le=365),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_staff),
):
    """Organization-wide risk analytics: distribution, trend, and department breakdown."""
    return get_risk_analytics(db, days=days)
